### PR TITLE
Add callback that performs adversarial training

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

Fixes #\<issue_number>

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] `pytest`
- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [ ] The title is **self-explanatory** and the description **concisely** explains the PR
- [ ] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [ ] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/callbacks/adversarial_training.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        input, target = batch
+
+        # FIXME: We reach into LitModular here...how can we get rid of this?
+        assert isinstance(pl_module, LitModular)
+        model = pl_module.model
+        sequence = model._sequences["training"]
+
+        # FIXME: This doesn't work because sequence does not include the Adversary module. How can we fix that?
+        #        Because this a callback, we can safely assume the Adversary module should live before the model.
+        #        We should be able to "manually" insert it into the sequence here.
```
Comment (dxoigmn) [inline mart/callbacks/adversarial_training.py:36]: Maybe we should make `mart.attack.Adversary` agnostic to the underlying model? That is, `Adversary` currently assumes the model it is attacking is a `LitModular`. Can we fix that?

Comment (dxoigmn) [reply mart/callbacks/adversarial_training.py:36]: It feels like there should be an `AdversaryLayer` (that is a `Module`) that wraps (or extends) an `Adversary` (that is `LightningModule`).

File: mart/configs/attack/adversary.yaml [hunk 2]
```diff
@@ -2,6 +2,7 @@ defaults:
   - /callbacks@callbacks: [progress_bar]
 
 _target_: mart.attack.Adversary
+_convert_: all
```
Comment (dxoigmn) [inline mart/configs/attack/adversary.yaml:5]: I feel like we should just convert everything in configuration by default. The use of `DictConfig` and `ListConfig` has tripped me up many times at this point.

**Comment by mzweilin** [comment]
Merged in https://github.com/IntelLabs/MART/pull/210

**Pull request closed without merging**